lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/avro/ProtocolSerializer.java

```
@@ -0,0 +1,196 @@
+    try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+      final String name = message.getClass().getSimpleName();
+      LOG.log(Level.INFO, "Serializing message: [" + name + "]");
+
+      final MessageSerializer serializer = nameToSerializerMap.get(name);
+      serializer.serialize(outputStream, message, identifier);
+
+      return outputStream.toByteArray();
+    } catch (Exception e) {
+      LOG.log(Level.SEVERE, "Failed to serialize avro message: " + e.getMessage());
+      return null;
+    }
+  }
+
+  /**
+   * Read a message from the input byte stream and send it to the event handler.
+   * @param messageBytes An array of bytes that contains the message to be deserialized.
+   * @param observer An implementation of the MultiObserver interface.
+   */
+  public void read(final byte[] messageBytes, final MultiObserver observer) {
+    try (InputStream inputStream = new ByteArrayInputStream(messageBytes)) {
+      // Binary decoder for both the header and the message.
+      final BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(inputStream, null);
+
+      // Read the header message.
+      final SpecificDatumReader<Header> headerReader = new SpecificDatumReader<>(Header.class);
+      final Header header = headerReader.read(null, decoder);
+      LOG.log(Level.INFO, "Deserializing message: [" + header.getClassName() + "]");
+
+      // Get the appropriate deserializer and deserialize the message.
+      final MessageDeserializer deserializer = nameToDeserializerMap.get(header.getClassName().toString());
+      if (deserializer != null) {
+        LOG.log(Level.INFO, "Calling message deserializer");
+        deserializer.deserialize(decoder, observer, header.getSequence());
+      } else {
+        LOG.log(Level.SEVERE, "Filed to find message deserializer");
+      }
+
+    } catch (Exception e) {
+      LOG.log(Level.SEVERE, "Failed to deserialize avro message: " + e.getMessage());
```

swallow the exception?